Adds sentry documentation and moves SENTRY_URL to SENTRY_DSN

anglinb opened the pull request:
👋 Error tracking is really important given [reporting errors to the user is hard](https://github.com/probot/friction/issues/1). However, we already have Raven bundled with probot so this just adds usage docs. 

This PR also deprecates the usage of environment variable `SENTRY_URL` in favor of `SENTRY_DSN`. I think it's nice to be consistent with most things that bundle in Raven:


<img width="1045" alt="client_keys" src="https://user-images.githubusercontent.com/2637602/29048015-6300e830-7b83-11e7-90ef-b6a8c09cf813.png">
However, if we think the trouble of deprecating this is too burdensome on developers, I can revert those changes. 

Fixes: https://github.com/probot/friction/issues/2

hiimbex commented:
👋 I read the docs and the friction issue, but I'm not really sure I understand what raven does. Any chance you could you provide me an example use case of what these changes would be and explanation? :D 

Also it would be ✨ if you could add a test to make sure the deprecation occurs like here: https://github.com/probot/probot/pull/193/files

bkeepers commented on docs/deployment.md:
```diff
@@ -119,3 +119,11 @@ To deploy a bot that includes multiple plugins, create a new app that has the pl
+
+## Error tracking
+
+Error tracking in production is of critical importance especially given that there is currently no recommended channel for letting the end user know that an error occurred. (We're working on a solution for end user error reporting [here](https://github.com/probot/friction/issues/1)). Probot comes bundled with [Raven](https://github.com/getsentry/raven-node) a client for the [Sentry](https://githu ...
```
> Error tracking in production is of critical importance especially given that there is currently no recommended channel for letting the end user know that an error occurred. (We're working on a solution for end user error reporting [here](https://github.com/probot/friction/issues/1)).

I feel like these could be removed and the docs would be just as helpful. These two sentences aren't wrong, they just aren't helpful.

bkeepers commented on index.js:
```diff
@@ -44,10 +44,16 @@ module.exports = (options = {}) => {
+  // Deprecate SENTRY_URL
+  if (process.env.SENTRY_URL && !process.env.SENTRY_DSN) {
+    process.env.SENTRY_DSN = process.env.SENTRY_URL;
+    console.warn('DEPRECATED: the `SENTRY_URL` key is now called `SENTRY_DSN`. Use of `SENTRY_URL` is deprecated and will be removed in 0.10.0');
```
> …will be removed in 0.10.0.

We should probably adopt an official deprecation policy, but recently I've been assuming that deprecated features will be around for one full minor version (while probot is <1.0, and one full major version once 1.0 is released). So a feature deprecated in 0.9 will still work in 0.10 and is removed in 0.11. This ensures that, as long as you don't currently have any deprecation warnings, you can always update to the next minor version without anything breaking.

bkeepers commented on lib/robot.js:
```diff
@@ -241,4 +241,9 @@ module.exports = (...args) => new Robot(...args);
+  * Note: Probot supports execption tracking through raven, a client for
+  * [sentry](https://github.com/getsentry/sentry). If the `SENTRY_DSN` is set
+  * as an environment varialbe, all errors will be forwarded to the sentry host
+  * specified by the environment variable.
```
> …all errors will be forwarded to the sentry host…

Is it any error passed to `robot.log`, or only those passed to `robot.log.error` (I haven't actually tested it)? Either way, might be good to specify that here.

anglinb replied:
I believe it forwards everything: https://github.com/transcovo/bunyan-sentry-stream/blob/1485adfff3df22db2e86f08a70fc2baddd336c02/lib/SentryStream.js#L33-L36

anglinb commented:
@bkeepers Will respond to your individual comments but I just wanted to touch on your question of a "meta" package. I think that's kind of a question for the project as a whole. I've seen project take on two different approaches: 1) Minimalist framework w/ easy plugins 2) Everything you need out of the box. The difference between these two is shown really in the difference between Django & Flask in the python ecosystem. 

To add my personal opinion, I think it should be as easy as possible to get started with probot but I think getting started in _the right way_ (ex: having error logging setup) is also worth adding complexity. I think a reasonable approach would be to split the difference use the Rails default gems model. Basically, we would add this "meta" package as a default dependency for all new probot plugins generated with the generate plugin command. Lmk your thoughts.

anglinb commented:
@hiimbex Yes, would love to explain this more! :D Sorry for the delay. (I will add this to docs but in the interim,) Sentry is an error tracking backend that acts as a resource for developers to track errors that get thrown in production. Raven is just a client to push information to Sentry. 

### An aside on error tracking:

We will never consider _every_ edge case and sometimes our applications will throw exceptions. At scale this is means there will possibly be a lot of exceptions, even if the error rates is 0.1% b/c there are a lot of requests. Additionally, in trying to reproduce an error, we want to know certain things about the application state like the payload that caused the exception, the current user, etc. A naive approach to error tracking would be to simply tail the logs that probot generates. However, this does not sufficiently address the two issue of exception volume and context necessary to reproduce the error. 

### Possible Solutions

There are a number of open source and paid services that address these issues. Bugsnag, Sentry, Yeller, etc. Sentry is one of the newest offering that addresses these issues well. It uses rollups to address the volume issue. A rollup is a single interface that shows an error class like crash on line x and all the instance of that error so a developer can quickly see what errors are happening a lot and can group individual errors next to other similar ones. Additionally Sentry provides "breadcrumbs" which track certain actions that succeeded before the request eventually errored out.  

I'm not sure why Sentry was adopted over the other solutions. Personally I really like sentry because its open source, their free tier is generous and they have really dope features on top of the two features most error trackers have, like deployment tagging and support for almost any language.

anglinb commented:
@hiimbex A couple quick questions: 1) Does that description make sense ^ ? 2) Do you think that level of detail makes sense for probot docs? A lot of what I explained is generally applicable to most projects so I'm not sure if it should live somewhere else and the probot docs should reference it or if I should inline in. Lmk your thoughts! 😄

anglinb commented on test/index.js:
```diff
@@ -97,4 +97,35 @@ describe('Probot', () => {
+        process.env.SENTRY_DSN = '1233';
+        expect(() => {
+          createProbot();
+        }).toThrow(/Invalid Sentry DSN: 1233/);
```
This feels a little hacky but I'm not sure what a better implementation would look like. It proves that Raven is getting instantiated with the `1233` value coming from the `SENTRY_DSN` variable. Happy to change the way these tests work given some feedback :D @bkeepers @hiimbex

bkeepers replied:
Seems fine to me.

hiimbex commented:
> 1) Does that description make sense ^ ? 

Yes, totally! Thanks :) Having not worked with that many huge projects before, exception handling hasn't really been something I've thought about often, but reading that and some of the sentry docs as well as seeing Brandon opening issues like this https://github.com/behaviorbot/welcome/issues/4 definitely helped my understanding and contextualized these changes.

> 2) Do you think that level of detail makes sense for probot docs? A lot of what I explained is generally applicable to most projects so I'm not sure if it should live somewhere else and the probot docs should reference it or if I should inline in.

Honestly, given my current understanding, I'm fine with the docs as is. I think that the kinds of developers who would want to use a platform like sentry probably have that base understanding, and if they don't, they probably just won't use it.
